Tree reductions for dask.array

jcrist opened the pull request:
This enables support for tree reductions, which should improve efficiency of using `dask.array` across multiple processes/machines, or when arrays are composed of a large number of chunks.

The idea is to set a maximum number of chunks to be gathered and combined (either overall, or by axis) when performing reductions - breaking the reduce step of map-reduce into a tree of smaller reductions.

At an api level, reductions expose a `max_leaves` kwarg, which defaults to the current behavior. It accepts either a `dict` of `{axis: max_chunks}`, or an integer, which is used to compute `max_chunks` for each dimension such that the total number of chunks gathered in each reduction is approximately `max_leaves`. For example, `axis=0, max_leaves=16 -> max_leaves={0: 16}`, `axis=(0, 1), max_leaves=16 -> max_leaves={0: 4, 1: 4}`.

Example:

```python
import dask.array as da
import numpy as np

x = np.arange(1, 122).reshape((11, 11)).astype('f4')
a = da.from_array(x, chunks=(4, 4))
o = a.sum(axis=0, max_leaves=4)
```
![mydask](https://cloud.githubusercontent.com/assets/2783717/11287254/d7860d8a-8edf-11e5-86a1-dbe05f9314a5.png)

What would normally be a single step reduction has been broken into a tree of depth 2.

Todo:
- [x] Add support for `moment` (and thus `std`, `var`, etc...)
- [x] Add support for `arg*` reductions
- [x] More tests

shoyer commented:
Does using a tree merge cause performance degradation for the single node case? If not, we might want to default to that behavior.

jcrist commented:
For a wide enough tree, no. It actually might improve performance, as the aggregate step can be done in parallel (probably negligible though). Trick is determining what is a sane default number.

mrocklin commented:
These produce different results

```python
In [11]: x.sum(max_leaves=4).visualize('dask.pdf')

In [12]: x.sum(max_leaves={0: 4}).visualize('dask.pdf')
```

mrocklin commented:
I wouldn't expect much degredation in non-pathological cases.  The worst thing we're doing is adding `O(log_k n)` more tasks.

Suggest renaming `max_leaves`.  I think that your intent of the name was something akin to `max_children`.  There might even be a slicker name than that.  I'm trying to think of a term like breadth that might be the converse of a depth of a tree.  Perhaps "branching factor", if it wasn't so verbose.

mrocklin commented on dask/array/reductions.py:
```diff
@@ -31,30 +34,73 @@ def reduction(x, chunk, aggregate, axis=None, keepdims=None, dtype=None):
-        return Array(dsk, result.name, chunks=chunks, dtype=dtype)
-    else:
-        return result
+    tmp = atop(partial(chunk, axis=axis, keepdims=True), inds, x, inds)
```
If you feel particularly underburdened we could move the `partial` call here into `atop`, allowing it to support keyword args.

jcrist replied:
Are you saying that `atop` should handle the partial itself?

mrocklin replied:
Perhaps

mrocklin commented on dask/array/reductions.py:
```diff
@@ -31,30 +34,73 @@ def reduction(x, chunk, aggregate, axis=None, keepdims=None, dtype=None):
+        tmp = partial_reduce(func, tmp, max_leaves, True, None)
+    func = compose(partial(aggregate, axis=axis, keepdims=keepdims),
+                   partial(_concatenate2, axes=axis))
+    return partial_reduce(func, tmp, max_leaves, keepdims, dtype)
```
Nice

mrocklin commented on dask/array/tests/test_reductions.py:
```diff
@@ -45,30 +45,35 @@ def reduction_1d_test(da_func, darr, np_func, narr, use_dtype=True):
 def test_reductions_1D_float():
     x = np.arange(5).astype('f4')
     a = da.from_array(x, chunks=(2,))
+    max_leaves = {0: 2}
```
It would be nice to verify that these work both with and without `max_leaves`.  This might be a good place for a for loop or pytest fixture.

jcrist replied:
All reductions are checked both with and without `max_leaves (although not for *all* cases), as seen in `reduction_2d_test`. A `pytest.mark.parametrize` would probably be cleaner though, rather than manually doing it.

mrocklin commented on dask/array/tests/test_reductions.py:
```diff
@@ -121,32 +128,44 @@ def reduction_2d_test(da_func, darr, np_func, narr, use_dtype=True):
 def test_reductions_2D_float():
     x = np.arange(1, 122).reshape((11, 11)).astype('f4')
     a = da.from_array(x, chunks=(4, 4))
+    max_leaves = {0: 2, 1: 2}
```
I know you haven't really filled out tests, so please excuse pedantic comments, but should also verify the accuracy of the branching as well for a few different kinds of branching.

jcrist replied:
All reductions are checked both with and without `max_leaves (although not for *all* cases), as seen in `reduction_2d_test`. A `pytest.mark.parametrize` would probably be cleaner though, rather than manually doing it.

mrocklin commented:
Any thoughts on how hard the `combine` function for moments will be?

mrocklin commented:
I haven't worked through the gritty details, but at first glance this looks pretty cool to me.

jcrist commented:
> These produce different results

What is the shape of x in that case? If it's not 1d, then I'd expect them too. I definitely need to clean up input verification though, as it occurs to me that I never verify that `sorted(axis) == sorted(max_leaves)`. Need max block combination sizes for all reduction axis.

mrocklin commented:
> What is the shape of x in that case? If it's not 1d, then I'd expect them to. I definitely need to clean up input verification though, as it occurs to me that I never verify that sorted(axis) == sorted(max_leaves). Need max block combination sizes for all reduction axis.

    x = da.arange(1000, chunks=100)

There is some input validation missing.  This is also the motivation behind the "we should test the branching factor explicitly somewhere" comment

jcrist commented:
I cleaned up the tests, and fixed the bug you pointed out above. Could always test more, but I think the coverage here is pretty good.

> Any thoughts on how hard the `combine` function for moments will be?

It's proving harder than I would like - my tired brain isn't up for mathing it out right now :/. Should be doable though, just need to do some math.

mrocklin commented on dask/array/tests/test_reductions.py:
```diff
@@ -287,3 +255,30 @@ def test_0d_array():
+    assert len(d2[1]) == 2
+    d3 = o.dask[d2[1][0]]
+    assert len(d3[1]) == 2
+    assert all(i[0].startswith('atop') for i in d3[1])
```
You could use `dask.core.get_deps` to get the `dependencies` dict and then use this dictionary in more sophisticated tests like the following:

```python
dependencies, dependents = get_deps(x.sum(max_leaves={0: 2, 1: 3}.dask)
assert max(map(len, dependencies.values()) == 2 * 3
```

There are presumably several such interesting configurations that would be useful to verify now, rather than several months from now when someone screws with this code unknowingly.

mrocklin replied:
Ping

jcrist replied:
Yeah, I saw that. Will fix.

mrocklin commented:
@jcrist what is the status on this?  I wouldn't mind using it to redo recent array experiments.

jcrist commented:
Doesn't support `arg_*`, `var`, `std`, or `moment`, and could use some more tests. I'm mostly out this week, so I doubt this will be fixed until next week. Should be good to go for redoing the array experiments though, if you don't mind working off non-merged stuff.

mrocklin commented:
I'm not sure if these should have the same name or not.

```python
In [1]: import dask.array as da

In [2]: x = da.random.random((100, 100), chunks=(10, 10))

In [3]: x.mean(axis=0).name
Out[3]: 'reduce-460ea5b3313d36450410978acc2ace95'

In [4]: x.mean(axis=0, max_leaves=6).name
Out[4]: 'reduce-460ea5b3313d36450410978acc2ace95'
```

mrocklin commented:
Using this branch I've found a speedup from 30s to 5s when doing a reduction of 2GB across a small network of 3 machines.  This is only when the reduction required data sharing and was only because we happened to have scattered the data in a way that aligns with this (both defaults cause nice behavior.)  There was no speedup or slowdown when the reduction didn't require significant data transfer, though that case was fast anyway (1-2s).

mrocklin commented:
Reasons to keep the keys the same: 

* They're mathematically the same result, just computed in different ways

Reasons to keep the keys separate and include `max_leaves` in tokenize:

*  The graphs are quite different, and so assumptions made about same keys having the same dependencies are incorrect.

jcrist commented:
I made them have the same keys so that it would play well with caching. I could revert this, but it seemed like the correct behavior in my mind. Where do we make assumptions that identical key names have the same dependencies?

mrocklin commented:
Within the distributed scheduler.

Or, more generally, within any scheduler that supports graph updates.

jcrist commented:
Ah, that makes sense. I don't have strong opinions on this, happy to go either way.

mrocklin commented:
My life will be easier if we include the branching factor within tokenize.  It also follows the "when in doubt, disambiguate keys" principle.

mrocklin commented:
I'm still in favor of changing the name away from `max_leaves`.  To me a leaf is strictly at the bottom of a tree.  This is a restriction both on leaves and on interior nodes.

jcrist commented:
Better name suggestion? Some ideas:
- `max_children`
- `max_branches`
- `branches`
- `branch_factor`
- `bfactor`

mrocklin commented:
I like `max_children` and `branch_factor`

mrocklin commented:
cc @shoyer, who tends to have a critical eye for APIs

jcrist commented:
While you're thinking about that, the default behavior still needs to be determined.

mrocklin commented:
I'm in favor of `max_leaves=8` until we find a counter-example

jcrist commented:
That seems small to me, but who knows. This will need more use to figure out best behavior.

mrocklin commented:
What are your thoughts on what it should be?

Also, we might want to make this a global managed by `set_options`.  I can imagine that most users won't want to manage it on every reduction call.

jcrist commented:
Not sure. I just want to avoid a deep tree composed of small chunks. If chunks are 10 MB, 8 chunks are tiny in memory. 64? 128? I don't really have a good handle on this yet. And yeah, it should be available in `set_options`.

mrocklin commented:
OK, lets say chunks are 10MB and we have 1TB of data.  That's 100,000 chunks.  

If we're doing a reduction then the result of the first reduction for each chunk is likely to be smaller, something between 1kB and 1MB.  We'll do something like a `ceil(log_8(100000)) = 6` deep reduction and have the following number of tasks

```python
>>> sum(100000 / 8**i for i in range(6))
114285.2783203125
```

If we switch this to 64 then we have

```python
In [6]: sum(100000 / 64**i for i in range(3))
Out[6]: 101586.9140625
```

Which does have significantly fewer extraneous tasks (it might make sense to weigh just communication bearing tasks, and not the original 100,000.)

This changes a bit if we consider nd-reductions, and which point our branching factor expands considerably.  

My experience here has been with image stacks where we end up doing our first stage reduction along a fairly shallow axis, e.g.

    >>> x.shape
    (small, big, big)
    >>> x.reduction(axis=0)

And our data ends up being spread by core.  It ends up being quite convenient to avoid communication on the second round of reductions.  I could also manage this by arranging data differently on the cluster however.

mrocklin commented:
64 would probably kill my small laptop on the weather data example.  

    720 lat * 1440 lon * 8 byte dtype * 64 blocks * 4 cores == 2GB

shoyer commented:
I would prefer not to need to pass through arguments to control the branching factor explicitly in reduction calls like `.mean()`. These sort of backend specific optimizations make it harder to write generic code. I haven't thought about this carefully enough yet to have specific suggestions on default behavior, though. Perhaps we could make some heuristic choices based on the operations and how the data is chunked?

mrocklin commented:
Heuristics sound like they would be valuable.  I think that we'll be better able to do this after some use.  I could just @jcrist 's help on some other things so I'm still behind a default that is just-a-number for now.

What are your thoughts on the kwarg names above?

shoyer commented:
I think `max_children` is better than `branch_factor`, which is more vague. It still feels more evocative for the particular implementation (using trees) rather than the concept (splitting the reduction into sub-problems).

More options:
`split_threshold`
`max_chunks`
`max_splits`
`reduce_limit`
`subproblem_size`
`max_subproblems`

I think `split_threshold` is my favorite.

jcrist commented:
This could use another review. I changed the keyword to `split_threshold`, added it to `set_options`, and defaulted at 32. Tests were improved, and all reductions are now supported.

mrocklin commented on dask/array/tests/test_reductions.py:
```diff
@@ -287,3 +269,51 @@ def test_0d_array():
+    assert_max_deps(x.sum(axis=(1, 2), split_threshold=20), 20, False)
+    assert_max_deps(x.sum(axis=(0, 1), split_threshold=40), 4 * 6)
+    assert_max_deps(x.sum(axis=(0, 2), split_threshold=40), 4 * 6)
+    assert_max_deps(x.sum(axis=(1, 2), split_threshold=40), 6 * 6)
```
This is a reassuring test.  Thanks!

mrocklin commented:
This looks great to me.

mrocklin commented:
+1

jcrist commented:
I expect this to need some use before we can determine a good default behavior/smart heuristic. I'm going to experiment with the ocean dataset to see how different configurations fair. If others have some workflows they could try this on, it would be much appreciated.

mrocklin commented:
I was playing with this with distributed with @stefanv .  Some issues came up

1.  The `split_threshold` kwarg wasn't immediately clear to him (perhaps he can chime in here on his thoughts
2.  We ran into an issue about reductions on flexible types while computing a mean.  This might be a distributed or a dask.array issue.  Some play testing is in order.  Flexible type issues arise when you try to call a typical reduction on an array of struct dtype.

jcrist commented:
There are no docs for this yet, so any suggestions on how to make this more intuitive to use would be much appreciated.

> We ran into an issue about reductions on flexible types while computing a mean.

I assume you weren't trying to compute a reduction on a struct array (which doesn't work even in numpy), but instead a bug somewhere in dask/distributed that resulted in the reduction being computed on a struct array? If you remember what computation caused it, it would be nice to be able to reproduce.

stefanv commented:
The reason ``split_threshold`` did not immediately make sense to me is because it contains these two concepts I needed to think about: ``split`` and ``threshold``.  And threshold is slightly confusing because it could imply "split after this condition" or "split until this condition", with potentially different meanings ("split after n chunks" or "split until you have n chunks").

So, while I don't propose this name, something like ``split_every_n_chunks`` would read very straightforwardly.  Maybe there's a shorter version of that, like ``chunks_per_split``, ``split_every``, ``group_chunks``, etc.

shoyer commented:
I like Stefan's names!

On Thu, Dec 3, 2015 at 7:51 PM, Stefan van der Walt
<notifications@github.com> wrote:

> The reason ``split_threshold`` did not immediately make sense to me is because it contains these two concepts I needed to think about: ``split`` and ``threshold``.  And threshold is slightly confusing because it could imply "split after this condition" or "split until this condition", with potentially different meanings ("split after n chunks" or "split until you have n chunks").
> So, while I don't propose this name, something like ``split_every_n_chunks`` would read very straightforwardly.  Maybe there's a shorter version of that, like ``chunks_per_split``, ``split_every``, ``group_chunks``, etc.
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/blaze/dask/pull/839#issuecomment-161866720

jcrist commented:
Sorry, this somehow slipped through. Of the new names, I like `split_every` the best, when used in a function call (i.e. `a.mean(axis=1, split_every=32)`). As a global/context config though with `set_options`, I like `split_ever_n_chunks` the best, as it's more descriptive. Unsure what's best here. Either way, a better name would be nice.

mrocklin commented:
I'd prefer using the same name in both places if possible.

On Fri, Dec 11, 2015 at 3:01 PM, Jim Crist <notifications@github.com> wrote:

> Sorry, this somehow slipped through. Of the new names, I like split_every
> the best, when used in a function call (i.e. a.mean(axis=1,
> split_every=32)). As a global/context config though with set_options, I
> like split_ever_n_chunks the best, as it's more descriptive. Unsure
> what's best here. Either way, a better name would be nice.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/blaze/dask/pull/839#issuecomment-164072350>.
>

jcrist commented:
Yeah, I wasn't suggesting different names in different spots. Perhaps just go with the more verbose one, as it will probably be set contextually/globally?

mrocklin commented:
I'd prefer `split_every` over `split_every_n_chunks`

shoyer commented:
I think split_every is probably verbose enough. I suspect it would usually
be poor practice to set this globally, given that it depends on the
particular reduction being performed.

On Fri, Dec 11, 2015 at 3:12 PM, Matthew Rocklin <notifications@github.com>
wrote:

> I'd prefer split_every over split_every_n_chunks
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/blaze/dask/pull/839#issuecomment-164073999>.
>

jcrist commented:
Done. See #876.
